circuit_passes/src/passes/mod.rs

```diff
@@ -37,19 +38,31 @@ pub trait CircuitTransformationPass {
     fn transform_circuit(&self, circuit: &Circuit) -> Circuit {
         self.pre_hook_circuit(&circuit);
         let templates = circuit.templates.iter().map(|t| self.transform_template(t)).collect();
-        let field_tracking = self.get_updated_field_constants();
+        let field_tracking = self.get_updated_field_constants(&circuit.llvm_data.field_tracking);
+        let bounded_loads = self.get_updated_bounded_array_loads(&circuit.llvm_data.bounded_array_loads);
+        let bounded_stores = self.get_updated_bounded_array_stores(&circuit.llvm_data.bounded_array_stores);
         let mut new_circuit = Circuit {
             wasm_producer: circuit.wasm_producer.clone(),
             c_producer: circuit.c_producer.clone(),
-            llvm_data: circuit.llvm_data.clone_with_new_field_tracking(field_tracking),
+            llvm_data: circuit.llvm_data.clone_with_updates(field_tracking, bounded_loads, bounded_stores),
             templates,
             functions: circuit.functions.iter().map(|f| self.transform_function(f)).collect(),
         };
         self.post_hook_circuit(&mut new_circuit);
         new_circuit
     }
 
-    fn get_updated_field_constants(&self) -> Vec<String>;
```

Comment:
I don't think we should use this new approach to getting updated field constants because it is very prone to losing updates. The `Value::to_value_bucket` function will add new constants to the `PassMemory` in some cases so it's kind of a hidden update that's easy to miss when writing passes. In fact, mapped_to_indexed, loop_unroll, and simplification all call that method at some point but you have removed the `get_updated_field_constants` implementation from those which means any updates made in those passes will be lost.

Reply:
Ah I see, I didn't catch this; I was just finding the boilerplate to be kinda clunky (i.e., all the passes have an effectively duplicated get field constants method). I'll revert.